File: code_examples/5-Holding_Data/1-Arrays/4-Array_2D/CPP/main.cpp
// Copyright (c) 2020 Mr. Coxall All rights reserved.
//
// Created by: Mr. Coxall
// Created on: Sep 2020
// This program uses a 2D array as a parameter

#include <iostream>
#include <cstdlib>
#include <ctime>


// In C++, a 2D array is passed by reference to a function
// (template is used to find the length of the array)
template <int rows, int cols>
int sumOfNumbers(int (&passedIn2DArray)[rows][cols]) {
    // this function adds up all of the numbers in a 2D array

    int total = 0;
    int counter = 0;

    //adding up all of the numbers in a 2D array
    for (int rowCounter = 0; rowCounter < rows; rowCounter++) {
        for (int columnCounter = 0; columnCounter < cols; columnCounter++) {
            total += passedIn2DArray[rowCounter][columnCounter];
        }
    }

    return total;
}

int main() {
    // this function uses a 2D array
    const int rows = 3;
    const int columns = 5;
    int a2DArray[rows][columns];
    int aSingleRandomNumber = 0;
    int sum = 0;

    srand(time(NULL));

    // input

    // So, ...
    //   In C++ you can't define array's size using variable.
    //   this is why you see const int rows = 2; above,
    //   so the size of the array can never change

    for (int rowElement = 0; rowElement < rows; rowElement++) {
        for (int columnElement = 0; columnElement < columns; columnElement++) {
            aSingleRandomNumber = (rand() % 9) + 1;
            a2DArray[rowElement][columnElement] = aSingleRandomNumber;
            std::cout << aSingleRandomNumber << " ";
        }
        std::cout << std::endl;
    }

    // call functions
    sum = sumOfNumbers(a2DArray);

    // output
    std::cout << "The sum of all the numbers is: " << sum << std::endl;

    std::cout << "\nDone." << std::endl;
    return 0;
}

File: code_examples/5-Holding_Data/1-Arrays/4-Array_2D/CSharp/main.cs
// Copyright (c) 2020 St. Mother Teresa HS All rights reserved.
//
// Created by: Mr. Coxall
// Created on: Sep 2020
// This program uses a 2D array

using System;

class Program
{
  static int SumOfNumbers(int[,] arrayOfNumbers)
  {
    // This function adds up all of the numbers in a 2D array

    int total = 0;
    int rows = arrayOfNumbers.GetLength(0);
    int columns = arrayOfNumbers.GetLength(1);

    // Loop through 2D array and add all numbers together
    for (int rowCounter = 0; rowCounter < rows; rowCounter++) {
      for (int columnCounter = 0; columnCounter < columns; columnCounter++) {
        total += arrayOfNumbers[rowCounter, columnCounter];
      }
    }

    return total;
  }

  static void Main()
  {
    // This function uses a 2D array
    const int rows = 7;
    const int columns = 5;
    int[,] number2DArray = new int[rows, columns];
    Random random = new Random();

    // Generate random numbers and populate the 2D array
    for (int row = 0; row < rows; row++)
    {
      for (int column = 0; column < columns; column++)
      {
        int aRandomNumber = random.Next(10);
        number2DArray[row, column] = aRandomNumber;
        Console.Write($"{aRandomNumber} ");
      }
      Console.WriteLine();
    }
    Console.WriteLine();

    // Call function
    int sumNumbers = SumOfNumbers(number2DArray);

    // Output
    Console.WriteLine($"The sum of all the numbers is: {sumNumbers}");

    Console.WriteLine("\nDone.");
  }
}
